action_server/src/action_server/test_tools.py

```diff
@@ -0,0 +1,152 @@
+    if the configuration step with the task manager raises an exception) are stored in a result dict. Eventually,
+    all positive results are printed and if negative results have occurred, an AssertionError is raised with a
+    message containing all failed options.
+
+    :param grammar: grammar that is used
+    :param grammar_target: top level grammar_target of the grammar
+    :raises: AssertionError if the test has failed
+    """
+    # Create variables
+    parser = CFGParser.fromstring(grammar)
+    task_manager = TaskManager(robot=Mockbot())
+    actions_rule = _find_actions_rule(parser, grammar_target)
+
+    test_results = {}
+    assert actions_rule.options, \
+        "There are no options in the actions rule. Therefore, there doesn't seem anything to test: that can't be right"
+    for option in actions_rule.options:  # type: Option
+
+        result_str = ""
+
+        # For each conjunct, we want to have a viable option
+        for conjunct in option.conjuncts:  # type: Conjunct
+            result_str += _resolve_conjunct(parser, conjunct.name)
+        result_str = result_str.rstrip(" ")  # type: str
+
+        # Now, we can parse the result string to get the action description
+        actions_definition = parser.parse_raw(
+            target=grammar_target,
+            words=result_str,
+        )  # type: dict
+        assert "actions" in actions_definition, "'actions' not in actions definition, don't know what to do"
+
+        # Test the action description with the task manager
+        try:
+            test_result = task_manager.set_up_state_machine(
+                recipe=actions_definition["actions"],
+            )  # type: ConfigurationResult
+        except Exception as e:
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            tb = "".join((traceback.format_exception(exc_type, exc_value, exc_traceback)))
+            test_result = ConfigurationResult()
+            test_result.message = "Configuration crashed: {}\n{}".format(e.message, tb)
+        test_results[result_str] = TMTestResult(actions_definition["actions"], test_result)
+
+    failed_test_results = {}
+    for action, test_result in test_results.iteritems():
+        if test_result.config_result.succeeded:
+            print("Configuration of '{}' succeeded".format(action))
+        elif test_result.config_result.message and test_result.config_result.missing_field:
+            print("Configuration of '{}' did not succeed due to missing information ({}): {}".format(
+                action, test_result.config_result.missing_field, test_result.config_result.message
+            ))
+        else:
+            failed_test_results[action] = test_result
+
+    error_str = "\n"
+    for action, test_result in failed_test_results.iteritems():
+        error_str += "\nConfiguration of action '{}' failed.\n\tRecipe: {}\n\tError: {}".format(
```

Comment:
Is is needed to start with a newline, when new items are also added with a new line?